### PR TITLE
[NFC][Clang] Use StringRef instead of string in ClangDiagnosticEmitter

### PR DESCRIPTION
**Pull request by jurahul**
Use StringRef instead of std::string in ClangDiagnosticEmitter.

### CODE REVIEW

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp [hunk 1]
```diff
@@ -158,21 +159,19 @@ static void groupDiagnostics(ArrayRef<const Record *> Diags,
...
   for (const Record *Group : DiagGroups) {
-    GroupInfo &GI =
-        DiagsInGroup[std::string(Group->getValueAsString("GroupName"))];
+    GroupInfo &GI = DiagsInGroup[Group->getValueAsString("GroupName")];
     GI.GroupName = Group->getName();
     GI.Defs.push_back(Group);
 
     for (const Record *SubGroup : Group->getValueAsListOfDefs("SubGroups"))
-      GI.SubGroups.push_back(SubGroup->getValueAsString("GroupName").str());
+      GI.SubGroups.push_back(SubGroup->getValueAsString("GroupName"));
```
Comment (jurahul) [inline clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:174]: Here's where the strings in the SubGroups vector originate from. As with most of my such changes, the strings are ultimately a part of the TableGen Records, which are ultimately backed by RecordKeeperImpl::StringInitStringPool. So the assumption is that the `RecordKeeper` (which is TableGen's IR) outlives the emitter backend invocation, which is true.

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp [hunk 2]
```diff
@@ -145,7 +146,7 @@ static bool diagGroupBeforeByName(const Record *LHS, const Record *RHS) {
          RHS->getValueAsString("GroupName");
 }
 
-using DiagsInGroupTy = std::map<std::string, GroupInfo, std::less<>>;
+using DiagsInGroupTy = std::map<StringRef, GroupInfo, std::less<>>;
```
[the remark this answers is not in the document]
Comment (jurahul) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:149]: Will do, I see that you added this just yesterday.

[the remark this answers is not in the document]
Comment (jurahul) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp]: Done.

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp [hunk 3]
```diff
@@ -117,15 +118,15 @@ namespace {
...
+    typedef std::vector<StringRef>::const_iterator const_iterator;
     const_iterator begin() const { return CategoryStrings.begin(); }
     const_iterator end() const { return CategoryStrings.end(); }
   };
 
   struct GroupInfo {
     StringRef GroupName;
     std::vector<const Record*> DiagsInGroup;
-    std::vector<std::string> SubGroups;
+    std::vector<StringRef> SubGroups;
```
[the remark this answers is not in the document]
Comment (kazutakahirata) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:129]: Thanks for the pointer and explanation!

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp (hunk 1 above)
Comment (kazutakahirata) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:174]: Thanks for the pointer and explanation!

**Comment by erichkeane** [comment]
FWIW, LGTM, I'm Ok with it as-is, but let @kazutakahirata do the approval when he's happy.

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp [hunk 4]
```diff
@@ -1419,16 +1412,14 @@ void clang::EmitClangDiagsDefs(const RecordKeeper &Records, raw_ostream &OS,
   InferPedantic inferPedantic(DGParentMap, Diags, DiagGroups, DiagsInGroup);
   inferPedantic.compute(&DiagsInPedantic, (RecordVec*)nullptr);
 
-  for (unsigned i = 0, e = Diags.size(); i != e; ++i) {
-    const Record &R = *Diags[i];
-
+  for (const Record *RP : Diags) {
+    const Record &R = *RP;
```
Comment (kazutakahirata) [inline clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:1416]: It's not ideal to have both `RP` and `R`.  You might want to clean this up in a follow-up patch or something.  Not a blocker for this patch.

**Review by kazutakahirata — APPROVED** [review]
LGTM.  Thanks!

File: clang/utils/TableGen/ClangDiagnosticsEmitter.cpp (hunk 4 above)
Comment (jurahul) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:1416]: I can clean this up in this PR. I was looking into whether LLVM has any iterator adapters for this earlier and could not find one. But looking again, it seems `make_pointee_range` does what we want, so let me try it

Comment (jurahul) [reply clang/utils/TableGen/ClangDiagnosticsEmitter.cpp]: Done.

**Pull request merged**